### DISCUSSION TITLE
trying to create 1-bit PNG

### DISCUSSION THREAD

**Discussion by Logicdemon** [ImageMagick/ImageMagick#7673] [Help]
### ImageMagick version

7.1.1-39

### Operating system, version and so on

Archlinux 6.11.2-arch1-1

### Description

I am looking for a way too convert a TIFF image into a PNG as losslessly as possible. However, every time I try a conversion to PNG, the file always seems to be up-converted to 8-bit Grayscale rather than staying as a 1-bit B&W image as in the original and also doubles the size! 

I've attached the TIFF as a zip file

[in.zip](https://github.com/user-attachments/files/17279312/in.zip)
  

I've tried many different combination of commands but seemingly can not get the PNG down to the same 1-bit colour of the TIFF file. I know I can increase the PNG compression settings, but I am just trying to make sure the correct colour format is being used first and not adding anything new to the PNG that was not in the TIFF. Below is just a snippet of some of the things I have tried. 

magick in.tif out.png			Try a basic command first. Comes out a 8-bit Grayscale
magick in.tif PNG00:out.png		I thought perhaps using PNG00 would inherit the right settings reading the in-file but came out again as 8-bit Grayscale
magick in.tif -type bilevel out.png	comes out as 8-bit Grayscale 
magick in.tif -depth 1 out.png	comes out as 8-bit Grayscale 

Running Identify look on both the TIFF and my outputs attempts I get :

identify * in.tif TIFF 2700x4701 2700x4701+0+0 1-bit Bilevel Gray 140477B 0.000u 0:00.000
identify * out.png PNG 2700x4701 2700x4701+0+0 8-bit Grayscale Gray 2c 295447B 0.000u 0:00.000

Running file I get:
file * in.tif: TIFF image data, little-endian, direntries=12, height=4701, compression=bi-level group 4, PhotometricInterpretation=WhiteIsZero, orientation=upper-left, width=2700
file * out.png: PNG image data, 2700 x 4701, 1-bit grayscale, non-interlaced

The TIFF file appears to be using compression=bi-level group 4. This is quite an old format, so I assumed PNG being newer format would at least be able rival or come close to the original file size. Is there a way to produce a 1-bit B&W PNG image losslessly, or is it simply a case of the compression used in the TIFF file is simply better than what can be done with PNG for this style of image? bi-level group 4 compression seems impressive, treating the black pixels like stencil I guess. but very annoying I can not replicate it in PNG.

I would be greateful for any insights on this, or just pointing out where I am obviously going wrong.

Cheers

**Comment by snibgo** [comment]
```
file * out.png: PNG image data, 2700 x 4701, 1-bit grayscale, non-interlaced
```
This tells you the file is "1-bit grayscale", which is what you want.

```
magick out.png -verbose info:
```
This gives you more information, if you want it:
```
  Class: PseudoClass
  Depth: 1-bit
  Colors: 2
  Histogram:
       2828859: (0,0,0) #000000 gray(0)
       9863841: (255,255,255) #FFFFFF gray(255)
```

We can check that the conversion is lossless:
```
magick compare -metric RMSE in.tif out.png NULL:
0 (0)
```
"0" means the images are identical.

Programs such as `pngcrush` can reduce the size (in bytes) of PNG files.

**Comment by jhcloos** [comment]
this:

> identify * out.png PNG 2700x4701 2700x4701+0+0 8-bit Grayscale Gray 2c 295447B 0.000u 0:00.000
> file * out.png: PNG image data, 2700 x 4701, 1-bit grayscale, non-interlaced

suggests that the misunderstanding is in reading identify's output.

the '2c' in the output says that there are only two colours in the png file.


i just confirmed that by running:

pbmmake 88 88 |pnmtopng >88x88.png
file 88x88.png
magick identify 88x88.png

and also got '1-bit grayscale' from file but '8-bit Grayscale Gray 2c'
from identify.

-JimC

**Comment by snibgo** [comment]
When `identify` says "`8-bit`", that means the `-depth` setting of the image is "8". We can change that, for example:
```
magick in.tif -depth 14 info:
in.tif TIFF 2700x4701 2700x4701+0+0 14-bit Gray 140477B 0.000u 0:00.000
```
The `-depth` setting has an effect on saving the image. It doesn't necessarily mean the input file has that many bits per channel.

**Comment by Logicdemon** [comment]
Thank you snibgo and jhcloos

The *compare* command is very useful! Its a bit of a rabbit hole too, a lot to read up on. Thank you for that. I was really hung up on the term *grayscale* each time I saw it. I was expecting just to see something like Black & white only and  assumed that meant the file was being up converted to 256 colour levels in there somewhere and my inital reaction leading me to think things were not lossless. 

Just wondering why does the command "identify out.png" list as 8-bit but "magick out.png -verbose info:" list as Depth: 1-bit? Or am I comparing the wrong things?

On compression. I am a little disappointed by the performance of PNG
 
in.tif			137.2 KB			original file
out.png		288.5 KB 		basic conversion (magick in.tif out-basic.png) no compression optimisation
outII.png		223.5 KB			png optimisation applied (ran through fileOptimizer)

Even with extensive optimisation the PNG file seems far larger compared to the TIFF... I can not get it any smaller than 223.5 KB

**Comment by snibgo** [comment]
"Grayscale" means there is only one channel, which records shades of gray. 

"-verbose info:" "Depth: 1-bit" examines the values of all the channels of all the pixels in the image, and calculates how many bits are needed to accurately record the values. When "-verbose" is not specified, IM doesn't do this calculation.

in.tif has 140477 bytes for 2700x4701 pixels, so that is 90 pixels per byte, or 11 pixels per bit. That is a good compression. Using PNG, the files is twice as large, so we still get 5.5 pixels per bit, which isn't bad.

If you want to minimize file sizes, different compression methods give different results for different images. Trying to predict the best method for a given image is not easy.

**Comment by fmw42** [comment]
If you want better compression for PNG, see pngcrush for example (or OptiPng) at https://imagemagick.org/Usage/formats/#png_non-im.  pngcrush tries many compression methods and values to try to optimize the compression. So it takes a while.  Also Imagemagick at https://imagemagick.org/script/command-line-options.php#quality for various values for -quality for PNG. Note that it is not the same as for JPG.

I think TIFF only has different compress methods ZIP and LZW and JPG for compressing. The first two are lossless and the last is lossy.

**Comment by Logicdemon** [comment]
>>>If you want to minimize file sizes, different compression methods give different results for different images. Trying to predict the best method for a given image is not easy.

I agree on that. 


>>>if you want better compression for PNG, see pngcrush for example (or OptiPng) 

I just run a bruteforce too. As far as I can tell I am using as good a chain of compression tools as I can. FileOptimizer is a wrapper that just moves files through a conveyor belt of compressors. Annoyingly, there is no linux CLI for this program (running it via wine). It can take a very long time for a PNG to pass through all the compressors but I don't think I do much better than going from a 288.5 KB png down to 223.5 KB. Here are some of the compressors it uses:

PngOptimizer 2.7 - http://psydk.org/PngOptimizer
PNGOUT 02/13/2015 - http://advsys.net/ken/utils.htm
pngquant 2.17.0 - http://www.pngquant.org
pngrewrite 1.4.0 - http://entropymine.com/jason/pngrewrite
pngwolf-zopli 1.1.2 - https://github.com/jibsen/pngwolf-zopfli
TruePNG 0.6.2.5 - http://x128.ho.ua/pngutils.html
OxiPNG 9.0.0 - https://github.com/shssoichiro/oxipng


Another test. I'm getting excellent results with JXL. However, currently JXL CLI does not support TIFF input, so have to use PNG as an intermediary exchange format.

in.tif			137.2 KB			original file
out.png		288.5 KB 		basic converstion (magick in.tif out-basic.png) no compression optimisation
outII.png		223.5 KB			png optimisation applied (ran through fileOptimizer)
outII.jxl		107.2 KB			( cjxl -v -d 0 -e 10 -E 11 -g 3 -I 100 outII.png outII.jxl )

JXL is not quite ready enough for me yet as encode parameters  are still being tweaked in the format, so I will be storing images as PNG until things stabilise a bit more. Hope to replace all the PNGs and JPGs I have stored as lossless JXL.
